dim/src/external/tmdb/mod.rs

```diff
@@ -0,0 +1,475 @@
+        let args: Vec<_> = args
+            .into_iter()
+            .map(|(k, v)| (k.to_string(), v.to_string()))
+            .collect();
+
+        let request = self.provider.http_client.get(url).query(&args);
+
+        async move {
+            let response = request
+                .send()
+                .await
+                .map_err(TMDBClientRequestError::reqwest)?;
+
+            let body = response
+                .bytes()
+                .await
+                .map_err(TMDBClientRequestError::reqwest)?;
+
+            std::str::from_utf8(&body)
+                .map_err(|_| TMDBClientRequestError::InvalidUTF8Body)
+                .map(|st| st.to_string())
+        }
+    }
+
+    async fn genre_list(&self, media_type: MediaType) -> Result<String, TMDBClientRequestError> {
+        self.make_request(
+            vec![("api_key", self.provider.api_key.as_ref())],
+            format!("/genre/{media_type}/list"),
+        )
+        .await
+    }
+
+    async fn search(
+        &self,
+        media_type: MediaType,
+        title: &str,
+        year: Option<i32>,
+    ) -> Result<String, TMDBClientRequestError> {
+        let mut args = vec![
+            ("api_key", self.provider.api_key.as_ref()),
+            ("language", "en-US"),
+            ("query", title),
+            ("page", "1"),
+            ("include_adult", "false"),
+        ]
+        .into_iter()
+        .map(|(k, v)| (k.to_string(), v.to_string()))
+        .chain(
+            year.into_iter()
+                .map(|n| ("year".to_string(), n.to_string())),
+        );
+
+        self.make_request(args, format!("/search/{media_type}"))
+            .await
+    }
+}
+
+// -- cache control
+
+/// The type of our hashmap we use for caching.
+///
+/// The current implementation is using [flurry](https://docs.rs/flurry)
+///
+type CacheMap = Arc<dashmap::DashMap<CacheKey, RwLock<Option<CacheValue>>>>;
+
+/// The key type used within the [CacheMap], refers to [CacheValue]s.
+///
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum CacheKey {
+    /// A search result
+    Search { title: String, year: Option<i32> },
+    /// Genre List
+    GenreList { media_type: MediaType },
+}
+
+type PendingRequestTx = broadcast::Sender<Result<Arc<str>, TMDBClientRequestError>>;
+
+/// The value type used within the [CacheMap], refered to by [CacheKey]s.
+#[derive(Clone)]
+enum CacheValue {
+    /// The request responsible for fulfilling this data is currently in flight.
+    RequestInFlight { tx: PendingRequestTx },
+    /// The responses body as UTF-8, cached.
+    Body { text: Arc<str> },
+}
+
+impl CacheValue {
+    /// get the data out of the value, if it is still pending, wait for it and turn errors into None.
+    async fn data(&self) -> Option<Arc<str>> {
+        match self {
+            CacheValue::RequestInFlight { tx } => {
+                tx.subscribe().recv().await.map(|o| o.ok()).ok().flatten()
```

Comment:
I think we should propagate the error to the caller, or at least log it.

nit: the map can just be a `.map(Result::ok)`.